### PR TITLE
Query routing with partition key definition

### PR DESCRIPTION
**Pull request by visridha**

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/CosmosQueryExecutionContextFactory.cs [hunk 1]
```diff
@@ -142,10 +143,31 @@ private async Task<CosmosQueryExecutionContext> CreateItemQueryExecutionContextA
             //todo:elasticcollections this may rely on information from collection cache which is outdated
             //if collection is deleted/created with same name.
             //need to make it not rely on information from collection cache.
+            PartitionKeyDefinition partitionKeyDefinition;
+            object partitionKeyDefinitionObject;
+            if (this.cosmosQueryContext.QueryRequestOptions != null
+                && this.cosmosQueryContext.QueryRequestOptions.Properties.TryGetValue(InternalPartitionKeyDefinitionProperty, out partitionKeyDefinitionObject))
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/CosmosQueryExecutionContextFactory.cs]: Is there any reason we don't just add this to CosmosQueryRequestOptions as a internal property?

Comment (visridha) [reply Microsoft.Azure.Cosmos/src/Query/CosmosQueryExecutionContextFactory.cs]: I don't want this to be long term - this is a temporary thing while we work out partition key routing in service interop.

File: Microsoft.Azure.Cosmos/src/Resource/QueryResponses/CosmosQueryResponse.cs [hunk 2]
```diff
@@ -10,6 +10,7 @@ namespace Microsoft.Azure.Cosmos
     using System.IO;
     using System.Linq;
     using System.Net;
+    using System.Runtime.CompilerServices;
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/QueryResponses/CosmosQueryResponse.cs]: Is it needed?

**Comment by kirankumarkolli** [comment]
Can you please add some more extra context on what the issues are. 
@j82w is it possible to UT these?

**Comment by j82w** [comment]
> Can you please add some more extra context on what the issues are.
> @j82w is it possible to UT these?

Unit tests are added

**Pull request merged**